README.md

gstatsd - A statsd service implementation in Python + gevent.

If you are unfamiliar with statsd, you can read [why statsd exists][etsy post],
or look at the [NodeJS statsd implementation][etsy repo].

License: [Apache 2.0][license]

Requirements
------------

 * [Python][python] - I'm testing on 2.6/2.7 at the moment.
 * [gevent][gevent] - A libevent wrapper.
 * [distribute][distribute] - (or setuptools) for builds.


Using gstatsd
-------------

Show gstatsd help:

    % gstatsd -h

Options:

    Usage: gstatsd [options] [config files]
    
     A statsd service in Python + gevent.
    
    Options:
      --version             show program's version number and exit
      -b BIND_ADDR, --bind=BIND_ADDR
                            bind [host]:port (host defaults to '')
      -s SINK, --sink=SINK  a service to which stats are sent
                            ([[host:]port:]type[,backend options]). Supported
                            types are "graphite" and "influxdb". InfluxDB backend
                            needs database, user, and password options, for
                            example: -s influxdb,mydb,myuser,mypass
      -v                    increase verbosity (currently used for debugging)
      -f INTERVAL, --flush=INTERVAL
                            flush interval, in seconds (default 10)
      -x KEY_PREFIX, --prefix=KEY_PREFIX
                            key prefix added to all keys (default None)
      -p PERCENT, --percent=PERCENT
                            percent threshold (default 90)
      -D, --daemonize       daemonize the service
      -h, --help 

Start gstatsd listening on the default port 8125, and send stats to graphite
server on its default port (2003) every 5 seconds:

    % gstatsd -s graphite -f 5

Bind listener to host 'foo' port 8126, and send stats to the Graphite server on
host 'bar' port 2003 every 20 seconds:

    % gstatsd -b foo:8126 -s bar:2003:graphite -f 20

To send the stats to multiple graphite servers, specify '-s' multiple times:

    % gstatsd -b :8125 -s stats1:2003:graphite -s stats2:2004:graphite

To send the stats influxdb server on its default port 8086 every 5 seconds:

    % gstatsd -s influxdb,mydb,myuser,mypass -f 5

To send the stats to both graphite and influxdb servers, specify '-s' multiple
times:

    % gstatsd -b :8125 -s stats1:2003:graphite -s stats2:8086:influxdb,mydb,myuser,mypass

You can also pass a yaml configuration file to gstatsd:

    % gstatsd ./config_example.yml

where 'config_example.yml' could be:

    host: localhost
    port: 8125
    daemonize: false
    verbose: false
    flush: 5
    prefix: 
    percent: 90
    sinks:
      - host: localhost
        port: 8086
        type: influxdb
        database: gstatsd
        user: gstatsd
        password: gstatsd
      - host: localhost
        port: 2003
        type: graphite


Using the client
----------------

The code example below demonstrates using the low-level client interface:

    from gstatsd import client

    # location of the statsd server
    hostport = ('localhost', 8125)

    raw = client.StatsClient(hostport)

    # add 1 to the 'foo' bucket
    raw.increment('foo')

    # timer 'bar' took 25ms to complete
    raw.timer('bar', 25)


You may prefer to use the stateful client:

    # wraps the raw client
    cli = client.Stats(raw)

    timer = cli.get_timer('foo')
    timer.start()

    ... do some work ..

    # when .stop() is called, the stat is sent to the server
    timer.stop()


[python]: http://www.python.org/
[gevent]: http://www.gevent.org/
[license]: http://www.apache.org/licenses/LICENSE-2.0
[distribute]: http://pypi.python.org/pypi/distribute
[etsy repo]: https://github.com/etsy/statsd
[etsy post]: http://codeascraft.etsy.com/2011/02/15/measure-anything-measure-everything/



debian/changelog
gstatsd (0.6+hl~12) wheezy; urgency=low

  * add support for using a uuid in the proxy keys

 -- Bruno Binet <bruno.binet@gmail.com>  Mon, 22 Sep 2014 16:56:16 +0200

gstatsd (0.6+hl~11) wheezy; urgency=low

  * add python-yaml as a dependency for gstatsd

 -- Bruno Binet <bruno.binet@gmail.com>  Mon, 22 Sep 2014 15:00:35 +0200

gstatsd (0.6+hl~10) wheezy; urgency=low

  * add support for date based rotation
  * quick&dirty error logging in /var/log/gstatsd/

 -- Bruno Binet <bruno.binet@gmail.com>  Mon, 22 Sep 2014 14:07:05 +0200

gstatsd (0.6+hl~9) wheezy; urgency=low

  * fix float conversion failure causing gevent thread crash

 -- Bruno Binet <bruno.binet@gmail.com>  Tue, 17 Jun 2014 11:18:41 +0200

gstatsd (0.6+hl~8) wheezy; urgency=low

  * add support for "gust" aggregation
  * fix bug: t value is wrong (1 interval ahead of time)
  * add support for filtering proxy with min_hour max_hour

 -- Bruno Binet <bruno.binet@gmail.com>  Fri, 13 Jun 2014 12:30:09 +0200

gstatsd (0.6+hl~7) wheezy; urgency=low

  * fix aggregate=last logic
  * make lastvaluefilesink try to make directory if missing

 -- Bruno Binet <bruno.binet@gmail.com>  Wed, 28 May 2014 13:12:43 +0200

gstatsd (0.6+hl~6) wheezy; urgency=low

  * add LastValueFileSink sink

 -- Bruno Binet <bruno.binet@gmail.com>  Mon, 19 May 2014 15:18:52 +0200

gstatsd (0.6+hl~5) wheezy; urgency=low

  * allow to enable/disable statsd numStats value through configuration

 -- Bruno Binet <bruno.binet@gmail.com>  Mon, 12 May 2014 12:37:35 +0200

gstatsd (0.6+hl~4) wheezy; urgency=low

  * send data to influxdb with microseconds precision
  * compute_timer_stats is not a method anymore
  * extract influxdb encoding logic in dedicated classmethod
  * extract graphite encoding logic in dedicated classmethod
  * writing numstats is now optional for graphite sink
  * update graphite sink to remove use of stringio
  * reuse GraphiteSink.encode method for the file sink
  * don't write statsd.numStats value when using the file sink
  * allow proxy key customization
  * change maxbytes default value to 4 Mo

 -- Bruno Binet <bruno.binet@gmail.com>  Fri, 11 Apr 2014 17:22:05 +0200

gstatsd (0.6+hl~3) wheezy; urgency=low

  * fix tests
  * split graphite and influxdb sinks into separate files
  * move tests to a dedicated 'test' subdirectory
  * add file sink support

 -- Bruno Binet <bruno.binet@gmail.com>  Thu, 27 Mar 2014 12:27:55 +0100

gstatsd (0.6+hl~2) wheezy; urgency=low

  * update debian build system
  * update .gitignore

 -- Bruno Binet <bruno.binet@gmail.com>  Wed, 05 Mar 2014 19:24:34 +0100

gstatsd (0.6+hl~1) wheezy; urgency=low

  * split received data on newlines
  * fix tests: _process signature has changed
  * generalize to support more than one sink type backend
  * free the memory buffer when we're done with it
  * factor out timer stats computation in a dedicated function
  * add support for influxdb backend
  * update optionparser doc for influxdb backend
  * fix _hosts and _urls attributes should be specfic to the instance
  * fix failing tests
  * update readmes
  * add support for reading options from a yaml config file
  * update class StatsConfig so that it accepts parameters
  * fix sinks attribute should be specfic to the instance
  * fix failing tests
  * add an example configuration file config_example.yml
  * add usage for optparse.OptionParser
  * update readmes to demonstrate the use of yaml configuration files
  * add support for proxy_values custom type
  * add support for proxy_values custom type for graphite backend
  * time precision is different for graphite vs influxdb
  * add support for proxy value to simple client lib
  * allow custom configuration for proxies (interval, aggregate)

 -- Bruno Binet <bruno.binet@gmail.com>  Mon, 03 Mar 2014 09:06:45 +0100

gstatsd (0.5) lucid; urgency=low

  * Release.

 -- Patrick Hensley <pat@clio>  Wed, 29 May 2013 12:44:32 -0400

gstatsd (0.4) lucid; urgency=low

  * Release.

 -- Patrick Hensley <spaceboy@indirect.com>  Wed, 29 Jun 2011 11:59:01 -0400

gstatsd (0.2) lucid; urgency=low

  * Initial release.

 -- Patrick Hensley <spaceboy@indirect.com>  Tue, 28 Jun 2011 21:10:19 -0400


gstatsd/config.py
import optparse
import re

from gstatsd.core import __version__


class ProxyConfig(object):

    def __init__(self, cfg, statscfg):
        self.name = cfg['name']
        self.regex = re.compile('^%s$' % self.name)
        self.allow = cfg.get('allow', True)
        self.interval = float(cfg.get('interval', statscfg.flush_interval))
        self.min_hour = int(cfg.get('min_hour', 0))
        self.max_hour = int(cfg.get('max_hour', 24))
        self.aggregate = cfg.get('aggregate', 'average')
        self.key = cfg.get('key', '%(uuid)s.%(hostname)s.%(name)s')
        assert self.allow in (True, False)
        assert self.aggregate in (
            'average', 'sum', 'min', 'max', 'last', 'gust')


class StatsConfig(object):
    props = ['uuid', 'host', 'port', 'verbose', 'flush_interval', 'prefix',
             'threshold', 'daemonize', 'numstats']
    # default config
    uuid = 'stats'
    host = 'localhost'
    port = 8125
    verbose = False
    flush_interval = 10
    prefix = None
    threshold = 90
    daemonize = False
    numstats = False

    def __init__(self, *args):
        self.sinks = []
        self.proxy = []
        for arg in args:
            if isinstance(arg, basestring):
                self.parse_yml(arg)
            elif isinstance(arg, dict):
                self.parse_dict(arg)
            elif isinstance(arg, optparse.Values):
                self.parse_options(arg)
            else:
                raise Exception('wrong type: %r' % arg)

    @classmethod
    def get_optionparser(cls):
        parser = optparse.OptionParser(
            description="A statsd service in Python + gevent.",
            usage="Usage: gstatsd [options] [config files]",
            version=__version__,
            add_help_option=False)
        parser.add_option(
            '-l', '--listen-host', dest='host',
            help="host to listen to (default '%s')" % cls.host)
        parser.add_option(
            '-p', '--listen-port', dest='port',
            help="port to listen to (default %d)" % cls.port)
        parser.add_option(
            '-s', '--sink', dest='sinks', action='append', default=[],
            help="a service to which stats are sent ([[host:]port:]type"
            "[,backend options]). Supported types are \"graphite\" and "
            "\"influxdb\".\nInfluxDB backend needs database, user, and "
            "password options, for example:\n-s influxdb,mydb,myuser,mypass")
        parser.add_option(
            '-f', '--flush-interval', dest='flush_interval',
            help="flush interval, in seconds (default %d)"
            % cls.flush_interval)
        parser.add_option(
            '-u', '--uuid', dest='uuid',
            help="uuid for this gstatsd instance (default %r)" % cls.uuid)
        parser.add_option(
            '-x', '--prefix', dest='prefix',
            help="key prefix added to all keys (default %r)" % cls.prefix)
        parser.add_option(
            '-t', '--threshold', dest='threshold',
            help="percent threshold (default %d)" % cls.threshold)
        parser.add_option(
            '-D', '--daemonize', dest='daemonize', action='store_true',
            help='daemonize the service')
        parser.add_option(
            '-n', '--numstats', dest='numstats', action='store_true',
            help='send statsd numstats value to all sinks')
        parser.add_option(
            '-v', '--verbose', dest='verbose', action='count',
            help="increase verbosity (currently used for debugging)")
        parser.add_option(
            '-h', '--help', dest='usage', action='store_true',
            help="display this message")
        return parser

    def parse_options(self, options):
        sinks = getattr(options, 'sinks', None)
        if sinks is not None:
            self.sinks += sinks
        for prop in self.props:
            val = getattr(options, prop, None)
            if val is not None:
                setattr(self, prop, val)

    def parse_yml(self, path):
        import yaml
        with open(path, 'r') as f:
            self.parse_dict(yaml.load(f))

    def parse_dict(self, d):
        if 'sinks' in d:
            self.sinks += d['sinks']
        if 'proxy' in d:
            self.proxy = [ProxyConfig(p, self) for p in d['proxy']]
        for prop in self.props:
            val = d.get(prop, None)
            if val is not None:
                setattr(self, prop, val)

    def dump_yml(self):
        import yaml
        d = {
            'sinks': self.sinks,
            }
        for prop in self.props:
            d[prop] = getattr(self, prop)
        return yaml.dump(d)
